# data_exploration 

In [20]:
import pandas as pd
import json

# Load the datasets
csv_data = pd.read_csv('data/raw data /Accidents-2019.csv')

with open('data/raw data /CalderdaleCollisions2020.json', 'r') as file:
    json_data = pd.json_normalize(json.load(file))

# Preview the datasets
print("CSV Data Preview:")
display(csv_data.head())

print("\nJSON Data Preview:")
display(json_data.head())


CSV Data Preview:


Reference Number  Grid Ref: Easting  Grid Ref: Northing  Number of Vehicles  \
0          6140231             398364              418232                   1   
1          6142035             407438              426457                   1   
2          6170291             409015              425004                   1   
3          6180214             409646              422315                   2   
4          6191021             407754              428825                   1   

   Accident Date  Time (24hr) 1st Road Class Road Surface  \
0          43469          701              A          Dry   
1          43469         2308   Unclassified          Dry   
2          43472          838   Unclassified          Dry   
3          43473          705              A          Dry   
4          43474         1515   Unclassified          Dry   

                       Lighting Conditions       Weather Conditions  ...  \
0             Darkness: no street lighting  Fine without high winds  ...   
1  Darkness: street lights present and lit  Fine without high winds  ...   
2          Daylight: street lights present  Fine without high winds  ...   
3          Daylight: street lights present  Fine without high winds  ...   
4          Daylight: street lights present  Fine without high winds  ...   

  Type of Vehicle 6 Casualty Class 6 Casualty Severity 6 Sex of Casualty 6  \
0               NaN              NaN                 NaN               NaN   
1               NaN              NaN                 NaN               NaN   
2               NaN              NaN                 NaN               NaN   
3               NaN              NaN                 NaN               NaN   
4               NaN              NaN                 NaN               NaN   

   Age of Casualty 6 Type of Vehicle 7 Casualty Class 7 Casualty Severity 7  \
0                NaN               NaN              NaN                 NaN   
1                NaN               NaN              NaN                 NaN   
2                NaN               NaN              NaN                 NaN   
3                NaN               NaN              NaN                 NaN   
4                NaN               NaN              NaN                 NaN   

  Sex of Casualty 7 Age of Casualty 7  
0               NaN               NaN  
1               NaN               NaN  
2               NaN               NaN  
3               NaN               NaN  
4               NaN               NaN  

[5 rows x 45 columns]


JSON Data Preview:


Reference Number  Grid Ref: Easting  Grid Ref: Northing  Number of vehicles  \
0          7151256             408276              424970                   2   
1          7160423             414597              421599                   1   
2          7171140             408932              426594                   1   
3          7171206             411284              420574                   1   
4          7180871             392922              423132                   1   

   Accident Date Time (24hr) 1st Road Class Road Surface  \
0          43835        1750   Unclassified   Wet / damp   
1          43836        1030              A          Dry   
2          43837        1640              A          Dry   
3          43837        1706              B          Dry   
4          43838        1441              A          Dry   

                       Lighting Conditions       Weather Conditions  ...  \
0  Darkness: street lights present and lit  Fine without high winds  ...   
1          Daylight: street lights present  Fine without high winds  ...   
2  Darkness: street lights present and lit     Fine with high winds  ...   
3  Darkness: street lights present and lit     Fine with high winds  ...   
4          Daylight: street lights present  Fine without high winds  ...   

  Type of Vehicle 5 Casualty Class 5 Casualty Severity 5 Sex of Casualty 5  \
0                                                                            
1                                                                            
2                                                                            
3                                                                            
4                                                                            

   Age of Casualty 5 Type of Vehicle 6 Casualty Class 6 Casualty Severity 6  \
0                                                                             
1                                                                             
2                                                                             
3                                                                             
4                                                                             

  Sex of Casualty 6 Age of Casualty 6  
0                                      
1                                      
2                                      
3                                      
4                                      

[5 rows x 40 columns]

In [18]:
print(csv_data.info())
print(json_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271 entries, 0 to 270
Data columns (total 45 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Reference Number     271 non-null    object
 1   Grid Ref: Easting    271 non-null    int64 
 2   Grid Ref: Northing   271 non-null    int64 
 3   Number of Vehicles   271 non-null    int64 
 4   Accident Date        271 non-null    int64 
 5   Time (24hr)          271 non-null    int64 
 6   1st Road Class       271 non-null    object
 7   Road Surface         271 non-null    object
 8   Lighting Conditions  271 non-null    object
 9   Weather Conditions   271 non-null    object
 10  Type of Vehicle 1    271 non-null    object
 11  Casualty Class 1     271 non-null    object
 12  Casualty Severity 1  271 non-null    object
 13  Sex of Casualty 1    271 non-null    object
 14  Age of Casualty 1    271 non-null    int64 
 15  Type of Vehicle 2    191 non-null    object
 16  Casualty

In [15]:
print("CSV Missing Data Summary:")
print(csv_data.isnull().sum())

print("\nJSON Missing Data Summary:")
print(json_data.isnull().sum())



CSV Missing Data Summary:
Reference Number         0
Grid Ref: Easting        0
Grid Ref: Northing       0
Number of Vehicles       0
Accident Date            0
Time (24hr)              0
1st Road Class           0
Road Surface             0
Lighting Conditions      0
Weather Conditions       0
Type of Vehicle 1        0
Casualty Class 1         0
Casualty Severity 1      0
Sex of Casualty 1        0
Age of Casualty 1        0
Type of Vehicle 2       80
Casualty Class 2        73
Casualty Severity 2     73
Sex of Casualty 2       73
Age of Casualty 2       74
Type of Vehicle 3      232
Casualty Class 3       221
Casualty Severity 3    221
Sex of Casualty 3      221
Age of Casualty 3      221
Type of Vehicle 4      256
Casualty Class 4       251
Casualty Severity 4    251
Sex of Casualty 4      251
Age of Casualty 4      251
Type of Vehicle 5      267
Casualty Class 5       265
Casualty Severity 5    265
Sex of Casualty 5      265
Age of Casualty 5      265
Type of Vehicle 6      268
Ca

In [16]:
print("Unique Road Surface Conditions in CSV:")
print(csv_data['Road Surface'].unique())

print("\nUnique Road Surface Conditions in JSON:")
print(json_data['Road Surface'].unique())


Unique Road Surface Conditions in CSV:
['Dry' 'Wet / Damp' 'Frost / Ice' 'Flood (surface water over 3cm deep)'
 'Snow']

Unique Road Surface Conditions in JSON:
['Wet / damp' 'Dry' 'Frost / ice' 'Snow' 'Wet  / damp' 'Wet /  damp']


# Data Exploration Summary

## CSV Dataset
1. **Shape**: 271 rows and 45 columns.
2. **Missing Values**:
   - Significant missing data in columns related to additional vehicles and casualties:
     - `Type of Vehicle 2`: 80 missing values.
     - `Type of Vehicle 3`: 232 missing values.
     - `Type of Vehicle 4-7`: Majority of rows are missing.
   - First vehicle and casualty details (`Type of Vehicle 1`, `Casualty Severity 1`, etc.) have no missing data.
3. **Unique Road Surface Conditions**:
   - Categories: `['Dry', 'Wet / Damp', 'Frost / Ice', 'Flood (surface water over 3cm deep)', 'Snow']`.


## JSON Dataset
1. **Shape**: 202 rows and 40 columns.
2. **Missing Values**:
   - Only the `Reference Number` column has 2 missing values.
   - All other columns have no missing data.
3. **Unique Road Surface Conditions**:
   - Categories:
     - Clean categories: `['Wet / damp', 'Dry', 'Frost / ice', 'Snow']`.
     - Duplicated formatting issues: `['Wet  / damp', 'Wet /  damp']`.


## Initial Observations
1. The JSON dataset is more complete but has formatting inconsistencies in `Road Surface`.
2. The CSV dataset has more missing values, especially for secondary vehicles and casualties.
3. Both datasets have some overlap in the structure (e.g., casualty and vehicle details) but may need cleaning for consistent formatting.


# data cleaning 
## Step 1: Planning the Cleaning Tasks

In [21]:
# Check for missing values
print("Missing Values in CSV Dataset:")
print(csv_data.isnull().sum())

print("\nMissing Values in JSON Dataset:")
print(json_data.isnull().sum())

# Check for duplicates
print("\nDuplicate Rows in CSV Dataset:", csv_data.duplicated().sum())
print("Duplicate Rows in JSON Dataset:", json_data.duplicated().sum())

# Check unique values for key columns
print("\nUnique Road Surface Conditions in CSV:")
print(csv_data['Road Surface'].unique())

print("\nUnique Road Surface Conditions in JSON:")
print(json_data['Road Surface'].unique())


Missing Values in CSV Dataset:
Reference Number         0
Grid Ref: Easting        0
Grid Ref: Northing       0
Number of Vehicles       0
Accident Date            0
Time (24hr)              0
1st Road Class           0
Road Surface             0
Lighting Conditions      0
Weather Conditions       0
Type of Vehicle 1        0
Casualty Class 1         0
Casualty Severity 1      0
Sex of Casualty 1        0
Age of Casualty 1        0
Type of Vehicle 2       80
Casualty Class 2        73
Casualty Severity 2     73
Sex of Casualty 2       73
Age of Casualty 2       74
Type of Vehicle 3      232
Casualty Class 3       221
Casualty Severity 3    221
Sex of Casualty 3      221
Age of Casualty 3      221
Type of Vehicle 4      256
Casualty Class 4       251
Casualty Severity 4    251
Sex of Casualty 4      251
Age of Casualty 4      251
Type of Vehicle 5      267
Casualty Class 5       265
Casualty Severity 5    265
Sex of Casualty 5      265
Age of Casualty 5      265
Type of Vehicle 6      2

## Step 2: Handle Missing Values


In [25]:
# Handle missing values in CSV dataset
# Drop columns with excessive missing data 
csv_data_cleaned = csv_data.drop(columns=[
    'Type of Vehicle 4', 'Casualty Class 4', 'Casualty Severity 4', 'Sex of Casualty 4', 'Age of Casualty 4',
    'Type of Vehicle 5', 'Casualty Class 5', 'Casualty Severity 5', 'Sex of Casualty 5', 'Age of Casualty 5',
    'Type of Vehicle 6', 'Casualty Class 6', 'Casualty Severity 6', 'Sex of Casualty 6', 'Age of Casualty 6',
    'Type of Vehicle 7', 'Casualty Class 7', 'Casualty Severity 7', 'Sex of Casualty 7', 'Age of Casualty 7'
])

# Replace missing values in Type of Vehicle 2 and 3 with 'Unknown'
csv_data_cleaned['Type of Vehicle 2'].fillna('Unknown', inplace=True)
csv_data_cleaned['Type of Vehicle 3'].fillna('Unknown', inplace=True)

# Handle missing values in JSON dataset
# Drop rows where Reference Number is missing
json_data_cleaned = json_data.dropna(subset=['Reference Number'])

# Preview cleaned datasets
print("CSV Data after handling missing values:")
print(csv_data_cleaned.info())

print("\nJSON Data after handling missing values:")
print(json_data_cleaned.info())



CSV Data after handling missing values:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271 entries, 0 to 270
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Reference Number     271 non-null    object
 1   Grid Ref: Easting    271 non-null    int64 
 2   Grid Ref: Northing   271 non-null    int64 
 3   Number of Vehicles   271 non-null    int64 
 4   Accident Date        271 non-null    int64 
 5   Time (24hr)          271 non-null    int64 
 6   1st Road Class       271 non-null    object
 7   Road Surface         271 non-null    object
 8   Lighting Conditions  271 non-null    object
 9   Weather Conditions   271 non-null    object
 10  Type of Vehicle 1    271 non-null    object
 11  Casualty Class 1     271 non-null    object
 12  Casualty Severity 1  271 non-null    object
 13  Sex of Casualty 1    271 non-null    object
 14  Age of Casualty 1    271 non-null    int64 
 15  Type of Vehicle 2

/var/folders/_8/jbknng3n4p90mytr00t30dsc0000gn/T/ipykernel_91826/3197937829.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  csv_data_cleaned['Type of Vehicle 2'].fillna('Unknown', inplace=True)
/var/folders/_8/jbknng3n4p90mytr00t30dsc0000gn/T/ipykernel_91826/3197937829.py:12: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we 

## Step 3: Standardize Formatting

In [27]:
# Standardize Road Surface in JSON dataset
json_data_cleaned['Road Surface'] = json_data_cleaned['Road Surface'].str.strip().replace({
    'Wet  / damp': 'Wet / Damp',
    'Wet /  damp': 'Wet / Damp',
    'Wet / damp': 'Wet / Damp'
})

# Preview unique values in Road Surface after standardization
print("Unique Road Surface Conditions in JSON (after cleaning):")
print(json_data_cleaned['Road Surface'].unique())


Unique Road Surface Conditions in JSON (after cleaning):
['Wet / Damp' 'Dry' 'Frost / ice' 'Snow']


/var/folders/_8/jbknng3n4p90mytr00t30dsc0000gn/T/ipykernel_91826/1846364798.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  json_data_cleaned['Road Surface'] = json_data_cleaned['Road Surface'].str.strip().replace({


## step 4 :Validate the Cleaned Data

In [28]:
# Check for duplicates in the cleaned datasets
print("Duplicate Rows in CSV Dataset (after cleaning):", csv_data_cleaned.duplicated().sum())
print("Duplicate Rows in JSON Dataset (after cleaning):", json_data_cleaned.duplicated().sum())

# Confirm data types in both datasets
print("\nCSV Dataset Data Types:")
print(csv_data_cleaned.dtypes)

print("\nJSON Dataset Data Types:")
print(json_data_cleaned.dtypes)


Duplicate Rows in CSV Dataset (after cleaning): 0
Duplicate Rows in JSON Dataset (after cleaning): 0

CSV Dataset Data Types:
Reference Number       object
Grid Ref: Easting       int64
Grid Ref: Northing      int64
Number of Vehicles      int64
Accident Date           int64
Time (24hr)             int64
1st Road Class         object
Road Surface           object
Lighting Conditions    object
Weather Conditions     object
Type of Vehicle 1      object
Casualty Class 1       object
Casualty Severity 1    object
Sex of Casualty 1      object
Age of Casualty 1       int64
Type of Vehicle 2      object
Casualty Class 2       object
Casualty Severity 2    object
Sex of Casualty 2      object
Age of Casualty 2      object
Type of Vehicle 3      object
Casualty Class 3       object
Casualty Severity 3    object
Sex of Casualty 3      object
Age of Casualty 3      object
dtype: object

JSON Dataset Data Types:
Reference Number       object
Grid Ref: Easting       int64
Grid Ref: Northing      

## step 5 : Save the Cleaned Datasets

In [32]:
# Save the cleaned CSV dataset
csv_data_cleaned.to_csv('data/cleaned data /Accidents-2019-cleaned.csv', index=False)

# Save the cleaned JSON dataset
json_data_cleaned.to_json('data/cleaned data /CalderdaleCollisions2020-cleaned.json', orient='records')

print("Cleaned datasets saved successfully!")

Cleaned datasets saved successfully!
